cmd/sniproxy/sniproxy.go

```diff
@@ -235,8 +366,41 @@ func (s *server) dnsResponse(req *dnsmessage.Message) (buf []byte, err error) {
 
 func (s *server) promoteHTTPS(ln net.Listener) {
 	err := http.Serve(ln, http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		httpPromoted.Add(1)
+		s.httpPromoted.Add(1)
 		http.Redirect(w, r, "https://"+r.Host+r.RequestURI, http.StatusFound)
 	}))
 	log.Fatalf("promoteHTTPS http.Serve: %v", err)
 }
+
+// initMetrics sets up local prometheus metrics, and creates clientmetrics to report those
+// same counters.
+func (s *server) initMetrics() {
+	s.stats = new(metrics.Set)
+
+	s.numTLSsessions = &expvar.Int{}
+	s.stats.Set("tls_sessions", s.numTLSsessions)
+	clientmetric.NewCounterFunc("sniproxy_tls_sessions", func() int64 { return s.numTLSsessions.Value() })
+
+	s.numTCPsessions = &metrics.LabelMap{Label: "proto"}
+	s.stats.Set("tcp_sessions", s.numTCPsessions)
+	// clientmetric doesn't have a good way to implement a Map type.
+	// We create clientmetrics dynamically when parsing the --forwards argument
+
+	s.numBadAddrPort = &expvar.Int{}
+	s.stats.Set("bad_addrport", s.numBadAddrPort)
+	clientmetric.NewCounterFunc("sniproxy_bad_addrport", func() int64 { return s.numBadAddrPort.Value() })
```

Comment:
I think this could more simply be:
```suggestion
	clientmetric.NewCounterFunc("sniproxy_bad_addrport", s.numBadAddrPort.Value)
```
and so on below (the `Value` method already has this signature)

Reply:
Done